### PR TITLE
Remove unnecessary parameter "found" from cxoCursor_fetchRow

### PR DESCRIPTION
**Pull request by alexhenrie**
This value can be safely encoded into the return value without having to pass around an extra pointer, and less function arguments means faster function calls.

### CODE REVIEW

File: src/cxoCursor.c [hunk 1]
```diff
@@ -1604,11 +1601,12 @@ static PyObject *cxoCursor_multiFetch(cxoCursor *cursor, int rowLimit)
 static PyObject *cxoCursor_fetchOne(cxoCursor *cursor, PyObject *args)
 {
     uint32_t bufferRowIndex = 0;
-    int found = 0;
```
Comment (cjbj) [inline src/cxoCursor.c:1607]: I'd prefer to keep this (and similar) initializations; not all static analyzers are smart.

Comment (cjbj) [reply src/cxoCursor.c:1607]: Ignore this!

**Comment by anthony-tuininga** [comment]
I understand that the `found` parameter can indeed be encoded in the return value of the function and there is some theoretical improvement in performance. That said, I suspect the difference is so negligible it isn't worth changing -- if you have a benchmark that suggests otherwise I'd love to see it! All of the functions return 0 for success and -1 for failure (if they return an integer) and changing this one doesn't seeem to make sense to me. I appreciate the earlier ones you sent but I'm not convinced this one is worth doing. Comments welcome!

**Comment by anthony-tuininga** [comment]
I understand that the `found` parameter can indeed be encoded in the return value of the function and there is some theoretical improvement in performance. That said, I suspect the difference is so negligible it isn't worth changing -- if you have a benchmark that suggests otherwise I'd love to see it! All of the functions return 0 for success and -1 for failure (if they return an integer) and changing this one doesn't seeem to make sense to me. I appreciate the earlier ones you sent but I'm not convinced this one is worth doing. Comments welcome!

**Comment by alexhenrie** [comment]
Here is a test program that runs a simple query 1000 times:

```
#!/usr/bin/env python3

import cx_Oracle
from neurtu import delayed, Benchmark

(
    host,
    port,
    sid,
    username,
    password,
    table
) = open('db-test.txt').read().split('\n')[:6]

dsn_tns = cx_Oracle.makedsn(host, int(port), sid)
dsn_tns = dsn_tns.replace('SID', 'SERVICE_NAME')
conn = cx_Oracle.connect(username, password, dsn_tns)
cursor = conn.cursor()

def grab_data():
    for row in cursor.execute('SELECT * FROM ' + table):
        pass

print(Benchmark(wall_time=True, cpu_time=True, repeat=1000)(delayed(grab_data)))
```

Database parameters are read from a file named `db-test.txt`.

With separate `found` variable (branch `master`):

```
$ ./test.py
         wall_time      cpu_time                                                                                              
mean  5.170811e-07  5.165712e-07
max   8.117774e-07  6.649632e-07
std   4.264257e-08  4.217068e-08
```

With `found` encoded into the function return value (branch `found`):

```
$ ./test.py
         wall_time      cpu_time
mean  5.030586e-07  5.023346e-07
max   6.095771e-07  5.909695e-07
std   3.918976e-08  4.006086e-08
```

There is about a 2% performance improvement:

```
$ python3 -c 'print(1-5.030586e-07/5.170811e-07)'
0.027118569988344388
```

And a statistical t-test shows that the result is highly significant (p<1e-13):

```
$ python3 -c 'from scipy.stats import ttest_ind_from_stats; print(ttest_ind_from_stats(5.170811e-07, 4.264257e-08, 1000, 5.030586e-07, 3.918976e-08, 1000).pvalue)'
2.955633741489284e-14
```

In addition to improving performance a bit, in my opinion the code is cleaner without the extra function argument. But I understand the desire to keep things consistent, so just use your best judgment here.

**Comment by alexhenrie** [comment]
It's also worth noting that many C functions use negative return values to report errors and positive return values to report status or progress. For example, [read](http://man7.org/linux/man-pages/man2/read.2.html) returns -1 on error, 0 on end-of-file, and 1 or more on success (depending on the number of bytes read). So distinguishing between a successful operation and "no more data" in the function return value is not an unusual design.

**Comment by anthony-tuininga** [comment]
I ran the test myself...and found that the results are highly variable. In fact, I found that the original code was about 0.5% faster in one run and then about 0.5% slower in the next run! Running it several times showed values ranging as much as 12%! And it didn't seem to matter which version was running, either. So I'm unconvinced that the difference you saw is highly significant. :-) Based on that result I'm inclined to leave the code unchanged -- consistency wins in this case, even though I would have willingly sacrificed it if the performance difference was indeed as noted.

**Comment by alexhenrie** [comment]
Well, that's why the test script did 1,000 runs: Clearly there is a lot of variance, but on average there was a 2% improvement by dropping the "found" parameter.

**Comment by anthony-tuininga** [comment]
Perhaps I wasn't clear! I did the 1000 runs about 7-8 times with each branch -- and discovered that the numbers varied among the batches of 1000 runs. Here are some samples:

With the original code:
```
         wall_time      cpu_time                                                
mean  4.725839e-07  4.732265e-07
max   5.183718e-07  5.559000e-07
std   6.221446e-09  6.812985e-09

         wall_time      cpu_time                                                
mean  4.122321e-07  4.108469e-07
max   5.395948e-07  4.980500e-07
std   8.336117e-09  7.672446e-09
```

With the modified code:
```
         wall_time      cpu_time                                                
mean  4.738659e-07  4.739669e-07
max   5.877571e-07  6.054500e-07
std   8.000417e-09  8.216466e-09

         wall_time      cpu_time                                                
mean  4.170724e-07  4.170669e-07
max   4.489678e-07  7.463000e-07
std   3.227856e-09  1.358115e-08
```

As you can see the numbers between batches are highly variable and (in this case) show that the modified code is actually worse! I don't believe that it actually is, though. Its just that it is impossible to tell the difference using code that requires round-trips to the database. And the other batches I ran showed similar numbers in between those two upper and lower bounds.

Is that clearer?

**Comment by alexhenrie** [comment]
Yeah that makes more sense. The results might be more clear on a CPU with a different architecture (e.g. a 32-bit CPU), but then again they might not be, so I can see why you wouldn't find this change compelling from a performance perspective.

**Comment by alexhenrie** [comment]
Thanks for looking at it with me!

**Pull request closed without merging**